src/codeu/chat/database/Table.java

```diff
@@ -35,4 +44,107 @@ public S getSchema() {
     return schema;
   }
 
+  /**
+   * Find objects with a given query.
+   *
+   * @param query The query.
+   * @param values The values for the query.
+   *
+   * @return The list of DBObjects found.
+   */
+  public List<DBObject<S>> findQuery(String query, String... values) {
+    Connection connection = database.getConnection();
+    if (connection == null) return new ArrayList<DBObject<S>>();
+
+    List<DBObject<S>> objects = new ArrayList<DBObject<S>>();
+
+    // Run a query in the database.
+    String sqlQuery = String.format("SELECT * FROM %s WHERE %s", name, query);
+    System.out.println(sqlQuery);
+    try (PreparedStatement stmt = connection.prepareStatement(sqlQuery)) {
+      int i = 1;
+      for (String value : values) {
+        stmt.setString(i ++, value);
+      }
+      try (ResultSet results = stmt.executeQuery()) {
+        // Try to find the first row.
+        while (results.next()) {
+          // Build a map of field values.
+          int id = -1;
+          Map<String, String> fields = new HashMap<String, String>();
+          for (String field : schema.getFields().keySet()) {
+            if (field.equals("_id")) {
+              id = results.getInt(field);
+            } else {
+              fields.put(name, results.getString(field));
+            }
+          }
+
+          // Create a DBObject with the field values.
+          DBObject<S> object = new DBObject<S>(this, id, fields);
+
+          // Add the object to the list.
+          objects.add(object);
+        }
+      }
+    } catch (SQLException ex) {
+      LOG.error("Failed to query database: ", ex.getMessage());
+    }
+
+    // Return the found objects.
+    return objects;
+  }
+
+  /**
+   * Find objects with given fields and values.
+   *
+   * @param parameters The field and value pairs.
+   *
+   * @return The list of DBObjects found.
+   */
+  public List<DBObject<S>> find(Map<String, String> parameters) {
+    // Build the query.
+    StringBuilder query = new StringBuilder();
+    int numFields = parameters.size();
+    for (String field : parameters.keySet()) {
+      query.append(String.format("%s = ?", field));
+      if (-- numFields > 0) {
+        query.append(" AND ");
+      }
+    }
+
+    return findQuery(query.toString(), parameters.values().toArray(new String[parameters.size()]));
+  }
+
+  /**
+   * Find objects with given fields and values.
+   *
+   * @param pairs The field and value pairs.
+   *
+   * @return The list of DBObjects found.
+   */
+  public List<DBObject<S>> find(String... pairs) {
```

Comment:
Please add a check to ensure the number of arguments is a multiple of 2, and throw a RuntimeException (like IllegalArgumentException) when that condition is not met.